### PR TITLE
148 internal operations [pump]

### PR DESCRIPTION
**Pull request by hleb-albau**

### CODE REVIEW

File: common/src/main/kotlin/fund/cyber/search/model/ethereum/OperationResult.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package fund.cyber.search.model.ethereum
+
+
+/**
+ * Represent result of operation invocation.
+ * For Reward and Destroy, result field always null.
+ */
+interface OperationResult
```
Comment (arturalbov) [inline common/src/main/kotlin/fund/cyber/search/model/ethereum/OperationResult.kt]: use sealed class

File: contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    ): List<EthereumContractSummaryDelta> {
+
+        val tx = record.value()
+        val fee = if (rootOperation) tx.fee else ZERO // fee applied only for tx -> only counted for root op
+        val opFailed = opResult is ErroredOperationResult
+        val balanceDelta = -(fee + if (opFailed) ZERO else op.value)
+
+        val contractDeltaByInput = EthereumContractSummaryDelta(
+            contract = op.from, lastOpTime = tx.blockTime!!, balanceDelta = balanceDelta,
+            successfulOpNumberDelta = if (rootOperation && opFailed) 0 else 1,
```
Comment (arturalbov) [inline contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt]: So, we count successfullOpNumberDelta if it's not rootOperation and it's not failed (rootOperation=false, opFailed=true). Is that correct?

File: contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            balanceDelta = if (opFailed) ZERO else op.value,
+            successfulOpNumberDelta = if (opFailed) 0 else 1,
+            txNumberDelta = if (isRootOp) 1 else 0, // tx count increased only for root op contracts
+            topic = record.topic(), partition = record.partition(), offset = record.offset()
+        )
+        return listOf(contractDeltaByInput, contractDeltaByOutput)
+    }
+
+    /**
+     * If parent operation failed, should no be invoked. Returns delta for destroyed contract and refund contract.
```
Comment (arturalbov) [inline contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt]: Is this Javadoc copied from `deltasByDestroyOperation()` method?

File: contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        return when (op) {
+            is CallOperation -> deltasByCallOperationTrace(op, opResult!!, isRootOp, record)
+            is CreateContractOperation -> deltasByCreateOperation(op, opResult!!, record, isRootOp)
+            is DestroyContractOperation -> deltasByDestroyOperation(op, record)
+            is RewardOperation -> emptyList() //not possible here
+        }
+    }
+
+    /**
+     * If parent operation failed, should no be invoked. Returns delta for destroyed contract and refund contract.
```
Comment (arturalbov) [inline contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt]: Is this Javadoc copied from `deltasByDestroyOperation()` method?

File: contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            txNumberDelta = if (rootOperation) 1 else 0, smartContract = false,
+            topic = record.topic(), partition = record.partition(), offset = record.offset()
+        )
+
+        if (opFailed) return listOf(contractDeltaByInput)
+
+        opResult as CreateContractOperationResult
+        val contractDeltaBySmartContract = EthereumContractSummaryDelta(
+            contract = opResult.address, lastOpTime = tx.blockTime!!, smartContract = true,
+            successfulOpNumberDelta = if (rootOperation && opFailed) 0 else 1,
```
Comment (arturalbov) [inline contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt]: The same as previous comment

File: contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        val deltasByOps = if (trace.rootOperationTrace.result !is ErroredOperationResult) {
+            rootOpsDeltas + deltasByOperations(trace.rootOperationTrace.subtraces, record)
+        } else {
+            rootOpsDeltas
+        }
+
+        return deltasByOps.map { delta -> if (event == PumpEvent.DROPPED_BLOCK) delta.revertedDelta() else delta }
+    }
+
+    override fun affectedContracts(records: List<ConsumerRecord<PumpEvent, EthereumTx>>): Set<String> {
```
Comment (arturalbov) [inline contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt]: It seems that you forgot to rewrite this method according to new logic.

File: contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt (hunk 2 above)
Comment (hleb-albau) [reply contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt]: Yes, method java doc says that this method invokes only for successful SUBops and root op. So For all subops  and only for not failed root op we increase counter

File: contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt (hunk 3 above)
Comment (hleb-albau) [reply contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt]: Yes, and it should be here.

Comment (hleb-albau) [reply contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt]: fix

File: contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt (hunk 4 above)
Comment (hleb-albau) [reply contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt]: fix

File: contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt (hunk 5 above)
Comment (hleb-albau) [reply contract-summary/ethereum/src/main/kotlin/fund/cyber/contract/ethereum/delta/EthereumTxDeltaProcessor.kt]: Yes, method java doc says that this method invokes only for successful SUBops and root op. So For all subops and only for not failed root op we increase counter

**Pull request merged**